source/common/common/inline_map_registry.h

```diff
@@ -0,0 +1,423 @@
+   */
+  uint64_t registrationNum() const {
+    ASSERT(finalized_, "Cannot fetch registration map before finalized()");
+    return registration_map_.size();
+  }
+
+  // Finalize this registry. No further changes are allowed after this point. This guaranteed that
+  // all map created by the process have the same variable size and custom registrations. This
+  // function could be called multiple times safely and only the first call will have effect.
+  void finalize() { finalized_ = true; }
+
+  template <class T> InlineMapPtr<T> createInlineMap() {
+    ASSERT(finalized_, "Cannot create inline map before finalized()");
+
+    // Call finalize() anyway to make sure that the registry is finalized and no any new inline
+    // keys could be registered.
+    finalize();
+    return InlineMap<T>::create(*this);
+  }
+
+#ifndef NDEBUG
+public:
+  // This is the registry id that the inline map created by. This is used to validate the inline
+  // key validity.
+  uint64_t registryId() const { return registry_id_; }
+
+private:
+  static uint64_t generateRegistryId() {
+    static std::atomic<uint64_t> next_id{0};
+    return next_id++;
+  }
+
+  uint64_t registry_id_{generateRegistryId()};
+#endif
+
+private:
+  bool finalized_{};
+  RegistrationSet registration_set_;
+  RegistrationMap registration_map_;
+};
+
+/**
+ * Helper class to get inline map registry singleton based on the scope. If cross-modules inline map
+ * registry is necessary, this class should be used.
+ */
+class InlineMapRegistryHelper {
+public:
+  using InlineKey = InlineMapRegistry<std::string>::InlineKey;
+  using ManagedRegistries = absl::flat_hash_map<std::string, InlineMapRegistry<std::string>*>;
+
+  /**
+   * Get the inline map registry singleton based on the scope.
+   */
+  template <class Scope>
+  static InlineMapRegistry<std::string>::InlineKey registerInlinKey(absl::string_view key) {
+    return mutableRegistry<Scope>().registerInlineKey(key);
+  }
+
+  /**
+   * Get registered inline key by the key name.
+   */
+  template <class Scope>
+  static absl::optional<InlineMapRegistry<std::string>::InlineKey>
+  getInlineKey(absl::string_view key) {
+    return mutableRegistry<Scope>().getInlineKey(key);
+  }
+
+  /**
+   * Create inline map based on the scope.
+   */
+  template <class Scope, class T>
+  static InlineMapRegistry<std::string>::InlineMapPtr<T> createInlineMap() {
+    return mutableRegistry<Scope>().template createInlineMap<T>();
+  }
+
+  /**
+   * Get all registries managed by this helper. This this helpful to finalize all registries and
+   * print the debug info of inline map registry.
+   */
+  static const ManagedRegistries& registries() { return mutableRegistries(); }
+
+  /**
+   * Finalize all registries managed by this helper. This should be called after all inline keys
+   * are registered.
+   */
+  static void finalize() {
+    for (auto& registry : mutableRegistries()) {
+      registry.second->finalize();
```

Comment:
heh -- I object to this function in principle. It assumes that a Central Scrutinizer knows about all the types of maps and when they are all ready to be initialized.

I think that's not reasonable, especially if the maps may be impacted by which filters are configured.

Reply:
1. This is function only impact the registries that created/managed by this helper and didn't impact other self-hosted registries.
2. `finalize()` could be called repeatly safely.
3. This will be called after boostrapping. This is additional constraints for registries created/managed by this helper. If user don't want this, just create and manage a registry by it self.